### ISSUE TITLE
EPIC: Usable Sprite Animation System (USAS) 

### ISSUE THREAD

**Issue by chairbender** [unitystation/unitystation#2311]
# Epic Description
This is an epic with multiple subtasks.

The goal of this epic is to create a convenient / usable system for defining, changing, and playing animations on a particular object , which includes non-animated / static sprites, and syncing them over the network. We also want this system to be flexible enough to support possible future use cases, such as having cool effects or particle emitters on clothes.

The current system uses SpriteHandler / SpriteDataHandler and while it does support various use cases we've encountered some difficulties trying to actually use and make sense of it.

Subtasks:
- [x] #2590 Simple World Icons
- [x] #2591 Composite World Icons
- [x] #2592 Simple Inventory Icons
- [x] #2593 Composite Inventory Icons
- [ ] #2594 Inhands
- [x] #2595 Clothing
- [x] #2596 Uniform Suit and Skirt Variants
- [x] #2597 Full Migration

Some guidance on the design and implementation:
- Try to get feedback early and often so you don't do a bunch of work that ends up having to be re-done.
- As far as flexibility is concerned (to support future or missed use cases), the general idea is that it should be set up so that we won't "code ourselves into a corner" and create a situation where lots of effort is needed to add capabilities to the system due to the design. Take a look at TileInteraction and its subclasses for an example of a somewhat flexible system. Inheritance / containment / interfaces can help here.
   - For example, in the future we might want to be able to apply a particular effect, shader, or material to something. Like a glowing effect. Or have it emit particles, such as a glittering crown.
- To help it coexist alongside the old system until it's completely migrated, you might want to make "V2" or "V3" versions of components you changed along with a common interface that the new / old components share (which would eventually be removed after migration) to make this easier.
- Re-evaluate the overall design of the system as we work on subtasks.
- Editor-assignable fields should almost always be private, lowercaseCamelCase, and use [SerializeField].
- Take care when naming classes / assets to maximize usability. Try to be precise but not too wordy.
- Be precise when naming fields. Longer field names are acceptable because they show up okay in the editor.
- Leverage scriptable objects. You can leverage inheritance and containment along with ScriptableObject classes to allow for lots of flexibility.
- Avoid using array/list fields where each index has a special meaning (like 0 is for inhands, 1 is for icon, etc...). This is difficult to make sense of. It's acceptable to use these fields for situations where the meaning of the order is obvious, such as defining a sequence of sprites to animate through.
- Try to use named fields rather than dictionary fields.
- Avoid using string fields, ESPECIALLY avoid using string dictionaries. Use enums and classes instead.
- The logic of the system doesn't all need to live in a single component or SO. I imagine this will involve changes to ItemAttributes (for defining in hands), and possibly a new Icon component for defining the appearance in inventory / world.

**Comment by chairbender** [comment]
IDK if I still believe the above design points I fleshed out.

I just really want to emphasize that EASE OF USE (for developers) is of the utmost importance.

**Comment by chairbender** [comment]
Rewrote the description to make it more clear so we can put a bounty on it.

**Comment by Bod9001** [comment]
I think using
https://docs.unity3d.com/Manual/com.unity.addressables.html
would help with 
A load times,
B build times,
C (Technically not an advantage of addressables but are present in newer versions of unity) with serialisation changes (as in list< list< list<ting>>>> Actually working) means that sprites could be stored ready to go 
D There could be a system to get the Addressable Asset ID?/Generating a ID for each addressable, then you can use that for networking what SO sprites you want to reference over the network?

all the load operations though will be asynchronous

**Comment by chairbender** [comment]
if that solves our networking pains trying to have networked IDs for SOs that would be great. That's been a big annoyance moving to SO-based stuff.

**Comment by Bod9001** [comment]
yeah, AddressableAssets is a no
reason:
cannot get key (the path for Loading assets) from unity reference ( Public assetreference bob ) so no Networking
but you can get GUID. so If you wanted to go with this approach you would have to make, Something that loads all the Stuff From the asset bundle, and Uses a identifying Things to synchronise across network, And work out a way of getting this for For mentioned "identifying Things" From a unity reference E.g ( Public assetreference bob

**Comment by DooblyNoobly** [comment]
Set bounty at B: 50 per request of a patron.

**Comment by Bod9001** [comment]
ok, With the asset bundles We would have to implement our own Editor script and loading up Sprites, Since they wouldn't be the unity references We are used to
So would have to make a Spritetoken
that When the game loads up it looks in the asset bundle for that token and uses the sprite from that

also we would also need a separate project for the sprites

**Comment by Bod9001** [comment]
looks like Addressable Asset has better editor integration so wouldn't have to make the UI to select what asset you would like to select, so  Addressable Asset  is a much better option if we are doing this

**Comment by Bod9001** [comment]
yeah, just saying I'm working on this currently

**Comment by Bod9001** [comment]
yeah, I'll pass the bounty on it putting back to pool, humm Actually let me think

**Comment by Bod9001** [comment]
I will just take 100 seems reasonable enough, I didn't do all the work only missing in hand stuff

**Comment by Bod9001** [comment]
Removing bounty seems kind of unnecessary
